File: src/main/java/data/FileInstance.java
package data;
import java.awt.Desktop;
import java.io.ByteArrayOutputStream;
import java.io.File;
import java.io.IOException;
import java.io.OutputStream;
import java.nio.file.Files;
import java.nio.file.Paths;
import java.text.DecimalFormat;
import java.util.ArrayList;
import java.util.Date;

public abstract class FileInstance {

	protected String name;
	protected final FileType type;
	protected final long size;
	protected final String path;
	protected Directory parent;

	public FileInstance(String path, String name, FileType type, Long size) {
		this.name = name;
		this.path = path;
		this.type = type != null ? type : FileType.UNK;
		this.size = size != null ? size : 0;
	}

	public String getName() { return this.name; }
	public FileType getType() { return this.type; }
	public long getSize() {	return this.size; }
	public String getPath() { return this.path; }
	public boolean isDirectory() { return this.type == FileType.DIR; }
	public boolean isEncryptable() { return false; }
	public Directory getParent() { return this.parent; }

	public String getFormattedSize() {
		if(size <= 0) return "0 B";
		final String[] units = new String[] { "B", "kB", "MB", "GB", "TB" };
		int digitGroups = (int) (Math.log10(size)/Math.log10(1024));
		return new DecimalFormat("#,##0.#").format(size/Math.pow(1024, digitGroups)) + " " + units[digitGroups];
	}

	public abstract ByteArrayOutputStream view() throws IOException;

	public void open() throws IOException {
		if(!Desktop.isDesktopSupported()) {  
			throw new UnsupportedOperationException("Opening files is not supported");
		}
		
		Desktop desktop = Desktop.getDesktop();  
		
		File file = new File(this.path);   
		if(file.exists()) {
			desktop.open(file);
		} else {
			throw new UnsupportedOperationException("Unable to open missing file");
		}
	}  
	
	public void openNew() throws IOException {
		if(!Desktop.isDesktopSupported()) {  
			throw new UnsupportedOperationException("Opening files is not supported");
		}
		
		Desktop desktop = Desktop.getDesktop();  
		
		String path = this.path.replace(this.name, "");
		String filename = null;
		if(this.name.contains(".")) {
			String[] parts = this.name.split("\\.");
			String ext = parts[parts.length-1];
			filename = this.name.replace("." + ext, "") + "-temp." + ext;
		} else {
			filename = this.name + "-temp";
		}
		
		Files.write(Paths.get(path + filename), view().toByteArray());
		
		File file = new File(path + filename);  
		if(file.exists()) {
			desktop.open(file);			
			// TODO may want to delete temp file
//			try { Thread.sleep(100); } catch (InterruptedException e) {}
//			Files.delete(Paths.get(file.getAbsolutePath()));
		} else {
			throw new UnsupportedOperationException("Unable to open missing file");
		}
	} 
	
	public void delete() throws IOException {
		Files.delete(Paths.get(this.path));
	}

	public void setName(String name) { this.name = name; }
	public void setParent(Directory parent) { this.parent = parent; }

	public String toString() { 
		return this.name + (this.type == FileType.DIR ? "/" : "") + 
				" (" + getFormattedSize() + ")"; 
	}




}

File: rs/src/data/file_instance.rs
use crate::data::file_type::FileType;
use std::collections::HashSet;
use std::fs::{self, File};
use std::io::{self, Read, Write};
use std::path::{Path, PathBuf};

struct FileInstance {
    name: String,
    file_type: FileType,
    size: usize,
    path: PathBuf,
    parent: Option<PathBuf>,
    tags: HashSet<String>,
}

impl FileInstance {
    pub fn new(path: &str, file_type: Option<FileType>, size: Option<usize>) -> Self {
        let path = Path::new(path);
        FileInstance {
            name: path.file_name().unwrap().to_str().unwrap().to_string(),
            file_type: file_type.unwrap_or(FileType::UNK),
            size: size.unwrap_or(0),
            path: PathBuf::from(path),
            parent: path.parent().and_then(|p| Some(PathBuf::from(p))),
            tags: HashSet::new(),
        }
    }

    pub fn is_directory(&self) -> bool {
        self.file_type == FileType::DIR
    }

    pub fn get_formatted_size(&self) -> String {
        if self.size == 0 {
            return "0 B".to_string();
        }

        let units = ["B", "kB", "MB", "GB", "TB"];
        let digit_groups = ((self.size as f64).log10() / (1024_f64).log10()) as i32;
        format!(
            "{:.1} {}",
            self.size as f64 / 1024_f64.powi(digit_groups),
            units[digit_groups as usize]
        )
    }

    fn view(&self) -> io::Result<Vec<u8>> {
        let mut file = File::open(&self.path)?;
        let mut buffer = Vec::new();
        file.read_to_end(&mut buffer)?;
        Ok(buffer)
    }

    fn open(&self) -> io::Result<()> {
        if cfg!(target_os = "windows") {
            std::process::Command::new("cmd")
                .args(&["/C", "start", &self.path.to_string_lossy()])
                .output()?;
        } else if cfg!(target_os = "macos") {
            std::process::Command::new("open")
                .arg(&self.path)
                .output()?;
        } else if cfg!(target_os = "linux") {
            std::process::Command::new("xdg-open")
                .arg(&self.path)
                .output()?;
        }
        Ok(())
    }

    fn open_new(&self) -> io::Result<()> {
        let path = self
            .path
            .parent()
            .unwrap_or_else(|| Path::new(""))
            .to_path_buf();
        let mut filename = self.name.clone();
        if let Some(extension) = self.path.extension() {
            filename = format!(
                "{}-temp.{}",
                self.path.file_stem().unwrap().to_string_lossy(),
                extension.to_string_lossy()
            );
        } else {
            filename.push_str("-temp");
        }

        let new_path = path.join(filename);
        fs::write(&new_path, self.view()?)?;

        // Open the file
        return self.open();
    }

    fn delete(&self) -> io::Result<()> {
        fs::remove_file(&self.path)
    }
}

impl ToString for FileInstance {
    fn to_string(&self) -> String {
        format!(
            "{}{} ({})",
            self.name,
            if self.is_directory() { "/" } else { "" },
            self.get_formatted_size()
        )
    }
}
